src/detect-filename.c

```diff
@@ -74,6 +74,124 @@ static int DetectEngineInspectFilename(
         const Signature *s,
         Flow *f, uint8_t flags, void *alstate, void *txv, uint64_t tx_id);
 
+DetectAppLayerParserKeywordEntry filename_list[] = {
+        {
+            .alproto = ALPROTO_HTTP, .directions = SIG_FLAG_TOCLIENT | SIG_FLAG_TOSERVER,
+            .inspect =  {
+                .Callback = DetectEngineInspectFilename
+            },
+            .mpm =  {
+                .priority = 2,
+                .PrefilterRegister = PrefilterMpmFilenameRegister
+            }
+        },
+        {
+            .alproto = ALPROTO_SMTP, .directions = SIG_FLAG_TOSERVER,
+            .inspect =  {
+                .Callback = DetectEngineInspectFilename
+            },
+            .mpm =  {
+                .priority = 2,
+                .PrefilterRegister = PrefilterMpmFilenameRegister
+            }
+        },
+        {
+            .alproto = ALPROTO_FTP, .directions = SIG_FLAG_TOCLIENT | SIG_FLAG_TOSERVER,
+            .inspect =  {
+                .Callback = DetectEngineInspectFilename
+            },
+            .mpm =  {
+                .priority = 2,
+                .PrefilterRegister = PrefilterMpmFilenameRegister
+            }
+        },
+        {
+            .alproto = ALPROTO_FTPDATA, .directions = SIG_FLAG_TOCLIENT | SIG_FLAG_TOSERVER,
+            .inspect =  {
+                .Callback = DetectEngineInspectFilename
+            },
+            .mpm =  {
+                .priority = 2,
+                .PrefilterRegister = PrefilterMpmFilenameRegister
+            }
+        },
+        {
+            .alproto = ALPROTO_SMB, .directions = SIG_FLAG_TOCLIENT | SIG_FLAG_TOSERVER,
+            .inspect =  {
+                .Callback = DetectEngineInspectFilename
+            },
+            .mpm =  {
+                .priority = 2,
+                .PrefilterRegister = PrefilterMpmFilenameRegister
+            }
+        },
+        {
+            .alproto = ALPROTO_NFS, .directions = SIG_FLAG_TOCLIENT,
+            .inspect =  {
+                .Callback = DetectEngineInspectFilename
+            },
+            .mpm =  {
+                .priority = 2,
+                .PrefilterRegister = PrefilterMpmFilenameRegister
+            }
+        }
+};
+
+DetectAppLayerParserKeywordEntry files_list[] = {
```

Comment:
maybe I'm overlooking something, but this seems to be a much more complex solution than needed.

Can't we just loop the existing app-layer registrations and just enable the keywords for each protocol that registers the `GetFiles` callback?

Reply:
The issue I found is that each file keyword handler usually has local/static functions for handling callbacks for some or all protocols -- thus, the need for local keyword data.

Beyond the local keyword data ... e.g., `files_list` from this source module, the remaining code is just ensuring those settings/registrations are made appropriately.

Reply:
I'm reworking the files storage into the transactions. I'm going to see if that affects this issue as well.